### DISCUSSION TITLE
required fields not being represented properly in the graphql schema

### DISCUSSION THREAD

**Discussion by chaoky** [payloadcms/payload#764] [General]
# Bug Report
required fields not being represented properly in the graphql schema

```ts
const Landing: GlobalConfig = {
  slug: "landing",
  versions: true,
  access: {
    read: (() => true )
  },
  fields: [
    {
      name: "title",
      type: "text",
      required: true,
      localized: true,
    },
  ],
}
```

## Expected Behavior
```graphql
type Landing {
  title: String!
}
```

## Current Behavior
```graphql
type Landing {
  title: String
}
```

**Comment by jmikrut** [comment]
Hey @chaoky — good catch!

This is actually working as intended because your field has `localized: true`. You can query documents in a locale where this field is not yet set, and no fallback locale, which means that the field may not have data present.

If you do not have `localized: true`, GraphQL will show this field as required properly.

Does this make sense? I am going to convert this into a discussion as I don't think there's much we can do here. Let me know what you think.

**Reply by chaoky** [reply]
this does make sense, thank you

**Reply by acomagu** [reply]
@jmikrut Are required fields non-null if fallback locale is set?

The current implementation does not seem to check if the fallback locale is set, can this be improved?

**Reply by jmikrut** [reply]
Well, regardless of `fallback` being enabled or not, the following is possible:

- Create a document in some locale
- Query in GraphQL, getting that document in one of your other locales 
- Boom, GraphQL gets pissed and everything goes bluey

This is why we have this type of thing in place. Of noting, the same thing applies to having any type of field-based access control, any type of conditional logic, etc. There are lots of cases in Payload where the field might be _**required**_ but not actually tell GraphQL or TS type generation that the field is required. 

Does that make sense?

**Reply by acomagu** [reply]
OK, I (actually tried it and) confirmed required&localized field can be null even if fallback is enabled. I didn't know that I could create records only in non-default locale.

So this make sense. Thank you!

**Comment by chaoky** [comment]
another possible issue is
```ts
{
  slug: "landing",
  fields: [
    {
      name: "cards",
      type: "blocks",
      blocks: [cards],
      required: true,
    },
  ],
}
```
outputting 

```
type Landing {
  cards: [Landing_Cards]
}
```
instead of 

```
type Foo {
  cards: [Landing_Cards!]!
}
```

**Reply by chaoky** [reply]
I could work on this or some way to work around it

**Reply by jmikrut** [reply]
Oh wow, good call! Thank you!

This could be a quick one. [Check here](https://github.com/payloadcms/payload/blob/master/src/graphql/schema/buildObjectType.ts#L455).

If you want to give it a shot, that would be great!

**Reply by chaoky** [reply]
small question https://github.com/payloadcms/payload/pull/799#issuecomment-1192939321
